### PR TITLE
feat: enable reward model training without dropping the last validation batch

### PR DESCRIPTION
**Pull request by jveronvialard**
# What does this PR do ?
This PR enables reward model training with validation_drop_last =False.

# Changelog 
- Please update the [CHANGELOG.md](/CHANGELOG.md) under next version with high level changes in this PR.

# Usage
```python
python ${NEMO_ALIGNER_PATH}/examples/nlp/gpt/train_reward_model.py \
    ...
    ++model.data.validation_drop_last=False \
    ...
```
# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation? Make sure to also update the [NeMo Framework User Guide](https://docs.nvidia.com/nemo-framework/user-guide/latest/index.html) which contains the tutorials

# Checklist when contributing a new algorithm
- [x] Does the trainer resume and restore model state all states?
- [x] Does the trainer support all parallelism techniques(PP, TP, DP)?
- [x] Does the trainer support `max_steps=-1` and `validation`?
- [x] Does the trainer only call APIs defined in [alignable_interface.py](/nemo_aligner/models/alignable_interface.py)?
- [x] Does the trainer have proper logging?

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py [hunk 1]
```diff
@@ -184,49 +184,57 @@ def gather_and_split_rewards(rewards_out):
...
+                # Compute accuracy over valid pairs only.
+                out_chosen, out_rejected = self.split_output_tensor(output_tensor)
+                comp = out_chosen > out_rejected
+                num_correct_pairs = comp[mask_valid_pairs].sum()
+                acc_chosen = num_correct_pairs / max(1, num_valid_pairs)
                 if validation_step and not self.cfg.data.get("validation_drop_last", True):
-                    num_valid_tokens_in_ub = batch["loss_mask"].sum()
+
                     if loss_for_ub.isnan():
                         assert batch["loss_mask"].count_nonzero() == 0, "Got NaN loss with non-empty input"
```
Comment (odelalleau) [inline nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py:201]: Could now be written
```suggestion
                        assert num_valid_pairs == 0, "Got NaN loss with non-empty input"
```

File: nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py [hunk 2]
```diff
@@ -184,49 +184,57 @@ def gather_and_split_rewards(rewards_out):
...
-
-                    return (
-                        loss_for_ub,
-                        {
-                            "loss_sum_and_ub_size": loss_sum_and_ub_size_all_gpu,
-                            "out_chosen": out_chosen,
-                            "out_rejected": out_rejected,
-                        },
-                    )
+                        reduced_acc = torch.zeros_like(num_correct_pairs)
```
Comment (odelalleau) [inline nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py:218]: Can you refactor those into a single `if`? (minor readability improvement)

File: nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py [hunk 3]
```diff
@@ -272,21 +278,21 @@ def get_loss_and_metrics(self, batch, forward_only):
...
-            acc_mean = acc_tensor.mean()
+            num_valid_pairs = torch.stack(
+                [loss_reduced["num_valid_pairs"] for loss_reduced in losses_reduced_per_micro_batch]
+            )
+            loss_tensor = torch.cat([loss_reduced["avg"].view(1) for loss_reduced in losses_reduced_per_micro_batch])
+            acc_tensor = torch.cat([loss_reduced["acc"].view(1) for loss_reduced in losses_reduced_per_micro_batch])
+
+            loss_mean = (loss_tensor * num_valid_pairs).sum() / num_valid_pairs.sum()
+            acc_mean = (acc_tensor * num_valid_pairs).sum() / num_valid_pairs.sum()
+            weights = num_valid_pairs.sum()
```
Comment (odelalleau) [inline nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py:289]: Minor rewrite but also casting `weights` to float as I suspect this may otherwise cause a hang with PP>1 when we broadcast the weights below:
```suggestion
            weights = num_valid_pairs.sum().float()
            loss_mean = (loss_tensor * num_valid_pairs).sum() / weights
            acc_mean = (acc_tensor * num_valid_pairs).sum() / weights
```

File: nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py [hunk 4]
```diff
@@ -272,21 +278,21 @@ def get_loss_and_metrics(self, batch, forward_only):
...
+
         else:
 
             loss_mean = torch.tensor(0.0, device=torch.cuda.current_device())
             acc_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            weights = torch.tensor(1.0, device=torch.cuda.current_device())
```
Comment (odelalleau) [inline nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py:295]: I feel like using 1 might be confusing since its value is irrelevant
```suggestion
            weights = torch.tensor(0.0, device=torch.cuda.current_device())
```

**Review by odelalleau — COMMENTED** [review]
Great, thanks! Just a few minor suggestions then should be good to go

File: nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py (hunk 3 above)
Comment (jveronvialard) [reply nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py]: Good point, thanks for flagging

**Review by odelalleau — APPROVED** [review]
Good to go, thanks!
